### PR TITLE
nxos_file_copy network_cli and httpapi fix

### LINKED ISSUE #40013: Intermittent Issue: Unable to copy large image files using nxos_file_copy module
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
nxos_file_copy

##### ANSIBLE VERSION
```
ansible 2.6.0 (rel260/nxos_upgrade 7bb4da05b6) last updated 2018/05/11 09:14:42 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /root/agents-ci/ansible/lib/ansible
  executable location = /root/agents-ci/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
```
DEFAULT_ROLES_PATH(env: ANSIBLE_ROLES_PATH) = [u'/root/agents-ci/ansible/test/integration/targets']
PARAMIKO_HOST_KEY_AUTO_ADD(env: ANSIBLE_PARAMIKO_HOST_KEY_AUTO_ADD) = True
PERSISTENT_COMMAND_TIMEOUT(env: ANSIBLE_PERSISTENT_COMMAND_TIMEOUT) = 600
PERSISTENT_CONNECT_TIMEOUT(env: ANSIBLE_PERSISTENT_CONNECT_TIMEOUT) = 600
```

##### OS / ENVIRONMENT
Ubuntu

##### SUMMARY
We used to be able to copy large image files using the `nxos_file_copy` module but now larger files fail to copy with the following error

```shell
  File "/usr/local/lib/python2.7/dist-packages/paramiko/transport.py", line 1610, in _send_user_message
    raise SSHException('Key-exchange timed out waiting for key negotiation')
```
##### STEPS TO REPRODUCE
Here are the image sizes of the files I am trying to copy

agent-lab9-pm:integration:2053> ls -lh /root/agents_images/
total 4.3G
-rw-r--r-- 1 root root  37M Mar  2 11:45 n3000-uk9-kickstart.6.0.2.U6.1a.bin
-rw-r--r-- 1 root root 845M Mar  2 18:42 nxos.7.0.3.I7.2.bin

```yaml
- name: "Copy {{ si }} to bootflash"
  nxos_file_copy:
    local_file: "{{image_dir}}{{ si }}"
    file_system: "bootflash:"
    provider: "{{ connection }}"
  register: result
```

##### EXPECTED RESULTS
Here is an example using a smaller image file
```shell
TASK [nxos_install_os : Copy n3000-uk9-kickstart.6.0.2.U6.1a.bin to bootflash] *******************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_install_os/tasks/upgrade/copy_kick_system_images.yaml:8
<n9k.example.com> connection transport is cli
<n9k.example.com> using connection plugin network_cli (was local)
<n9k.example.com> starting connection from persistent connection plugin
<n9k.example.com> found existing local domain socket, using it!
<n9k.example.com> updating play_context for connection
<n9k.example.com> 
<n9k.example.com> local domain socket path is /root/.ansible/pc/fe6d223170
<n9k.example.com> socket_path: /root/.ansible/pc/fe6d223170
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_file_copy.py
<n9k.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n9k.example.com> EXEC /bin/sh -c 'echo ~root && sleep 0'
<n9k.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313 `" && echo ansible-tmp-1526046585.58-51328088639313="` echo /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313 `" ) && sleep 0'
<n9k.example.com> PUT /root/.ansible/tmp/ansible-local-31112XrsRA_/tmpPn8qX7 TO /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313/nxos_file_copy.py
<n9k.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313/ /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313/nxos_file_copy.py && sleep 0'
<n9k.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313/nxos_file_copy.py && sleep 0'
<n9k.example.com> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1526046585.58-51328088639313/ > /dev/null 2>&1 && sleep 0'
changed: [n9k.example.com] => {
    "changed": true, 
    "file_system": "bootflash:", 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": null, 
            "connect_ssh_port": 22, 
            "file_system": "bootflash:", 
            "host": "n9k.example.com", 
            "local_file": "/root/agents_images/n3000-uk9-kickstart.6.0.2.U6.1a.bin", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": null, 
            "provider": {
                "auth_pass": null, 
                "authorize": null, 
                "host": "n9k.example.com", 
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "port": null, 
                "ssh_keyfile": null, 
                "timeout": null, 
                "transport": "cli", 
                "use_proxy": true, 
                "use_ssl": null, 
                "username": "admin", 
                "validate_certs": null
            }, 
            "remote_file": null, 
            "ssh_keyfile": null, 
            "timeout": null, 
            "transport": "cli", 
            "use_proxy": true, 
            "use_ssl": null, 
            "username": "admin", 
            "validate_certs": null
        }
    }, 
    "local_file": "/root/agents_images/n3000-uk9-kickstart.6.0.2.U6.1a.bin", 
    "remote_file": "n3000-uk9-kickstart.6.0.2.U6.1a.bin", 
    "transfer_status": "Sent"
}
```
##### ACTUAL RESULTS
Here is an example using a larger image file
```
TASK [nxos_install_os : Copy nxos.7.0.3.I7.2.bin to bootflash] ***********************************************************************************************************************************
task path: /root/agents-ci/ansible/test/integration/targets/nxos_install_os/tasks/upgrade/copy_kick_system_images.yaml:8
<n9k.example.com> connection transport is cli
<n9k.example.com> using connection plugin network_cli (was local)
<n9k.example.com> starting connection from persistent connection plugin
<n9k.example.com> found existing local domain socket, using it!
<n9k.example.com> updating play_context for connection
<n9k.example.com> 
<n9k.example.com> local domain socket path is /root/.ansible/pc/ee1ab4a788
<n9k.example.com> socket_path: /root/.ansible/pc/ee1ab4a788
Using module file /root/agents-ci/ansible/lib/ansible/modules/network/nxos/nxos_file_copy.py
<n9k.example.com> ESTABLISH LOCAL CONNECTION FOR USER: root
<n9k.example.com> EXEC /bin/sh -c 'echo ~root && sleep 0'
<n9k.example.com> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141 `" && echo ansible-tmp-1526046216.79-32071527780141="` echo /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141 `" ) && sleep 0'
<n9k.example.com> PUT /root/.ansible/tmp/ansible-local-30569sYjrvT/tmpbMF3OR TO /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141/nxos_file_copy.py
<n9k.example.com> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141/ /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141/nxos_file_copy.py && sleep 0'
<n9k.example.com> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141/nxos_file_copy.py && sleep 0'
<n9k.example.com> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1526046216.79-32071527780141/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_HFaPp8/ansible_module_nxos_file_copy.py", line 174, in transfer_file
    scp.put(module.params['local_file'], full_remote_path)
  File "/usr/local/lib/python2.7/dist-packages/scp.py", line 154, in put
    self._send_files(files)
  File "/usr/local/lib/python2.7/dist-packages/scp.py", line 249, in _send_files
    chan.sendall(file_hdl.read(buff_size))
  File "/usr/local/lib/python2.7/dist-packages/paramiko/channel.py", line 813, in sendall
    sent = self.send(s)
  File "/usr/local/lib/python2.7/dist-packages/paramiko/channel.py", line 767, in send
    return self._send(s, m)
  File "/usr/local/lib/python2.7/dist-packages/paramiko/channel.py", line 1144, in _send
    self.transport._send_user_message(m)
  File "/usr/local/lib/python2.7/dist-packages/paramiko/transport.py", line 1610, in _send_user_message
    raise SSHException('Key-exchange timed out waiting for key negotiation')

fatal: [n9k.example.com]: FAILED! => {
    "changed": false, 
    "file_size": 885554688, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": null, 
            "connect_ssh_port": 22, 
            "file_system": "bootflash:", 
            "host": "n9k.example.com", 
            "local_file": "/root/agents_images/nxos.7.0.3.I7.2.bin", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": null, 
            "provider": {
                "auth_pass": null, 
                "authorize": null, 
                "host": "n9k.example.com", 
                "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
                "port": null, 
                "ssh_keyfile": null, 
                "timeout": null, 
                "transport": "cli", 
                "use_proxy": true, 
                "use_ssl": null, 
                "username": "admin", 
                "validate_certs": null
            }, 
            "remote_file": null, 
            "ssh_keyfile": null, 
            "timeout": null, 
            "transport": "cli", 
            "use_proxy": true, 
            "use_ssl": null, 
            "username": "admin", 
            "validate_certs": null
        }
    }, 
    "msg": "Could not transfer file. There was an error during transfer. Please make sure remote permissions are set.", 
    "temp_size": "534495232"
}
	to retry, use: --limit @/root/agents-ci/ansible/test/integration/nxos.retry

PLAY RECAP ***************************************************************************************************************************************************************************************
n9k.example.com          : ok=41   changed=1    unreachable=0    failed=1 
```

### PR DESCRIPTION
**Pull request by trishnaguha**
##### SUMMARY

fixes https://github.com/ansible/ansible/issues/40013#issuecomment-391129484

- Pass play_context host, user, password to the module code.
- Remove provider

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

modules/network/nxos/nxos_file_copy.py
plugins/action/nxos.py
test/integration/targets/nxos_file_copy/
##### ANSIBLE VERSION

```
devel
```

**Result**
All the tests passed.
```
PLAY RECAP ********************************************************************************************************************************************************************************************************
nxos9k-02                  : ok=84   changed=12   unreachable=0    failed=0   
```

### CODE REVIEW

**Comment by Qalthos** [comment]
Okay, I see what the deal is now. My only suggestion is to blacklist this module from running over nxapi (like we do for nxos_nxapi) as it is at best misleading

**Comment by mikewiebe** [comment]
@Qalthos Agree that since this module uses there is no reason to test this over `nxapi` since nxapi is never involved.

**Comment by trishnaguha** [comment]
@Qalthos https://github.com/ansible/ansible/pull/40592/commits/9e7bfa9ae7d769583cf486d8cac33f7bd99465bc removes provider.

File: lib/ansible/plugins/action/nxos.py [hunk 1]
```diff
@@ -43,8 +43,18 @@ def run(self, tmp=None, task_vars=None):
 
         socket_path = None
 
+        if self._task.action == 'nxos_file_copy':
+            if self._play_context.connection == 'network_cli':
```
Comment (Qalthos) [inline lib/ansible/plugins/action/nxos.py]: `host` and `password` don't change, I would just set them all as if using network_cli, then `if connection == local: username = connection_user`. This has the side benefit of being really easy to clean up when local goes away

Comment (trishnaguha) [reply lib/ansible/plugins/action/nxos.py]: @Qalthos Of course. I just did that to keep the blocks distinct. But I don't have strong opinion on this. Makes sense and pushed.

**Review by Qalthos — APPROVED** [review]
I mentioned elsewhere what I wanted was just setting the variables as if network_cli, and then resetting username if we're in local, but this is fine, too

**Pull request merged**